src/internal_telemetry/allocations/allocator/tracing_allocator.rs

```diff
@@ -53,40 +51,78 @@ unsafe impl<A: GlobalAlloc, T: Tracer> GlobalAlloc for GroupedTraceableAllocator
         let (group_id_ptr, object_ptr, _wrapped_layout) =
             self.get_wrapped_allocation(object_layout);
         let object_size = object_layout.size();
-
-        try_with_suspended_allocation_group(
-            #[inline(always)]
-            |group_id| {
-                group_id_ptr.write(group_id.as_raw());
-                self.tracer.trace_allocation(object_size, group_id);
-            },
-        );
+        // Group id value of zero implies allocations tracking was disabled
+        // during this allocation. We override this if allocations were in fact enabled.
+        group_id_ptr.write(0);
+        if TRACK_ALLOCATIONS.load(Ordering::Relaxed) {
+            try_with_suspended_allocation_group(
+                #[inline(always)]
+                |group_id| {
+                    group_id_ptr.write(group_id.as_raw());
+                    self.tracer.trace_allocation(object_size, group_id);
+                },
+            );
+        }
 
         object_ptr
     }
 
+    #[track_caller]
+    unsafe fn alloc_zeroed(&self, object_layout: Layout) -> *mut u8 {
+        // Allocate our wrapped layout and make sure the allocation succeeded.
+        let (actual_layout, offset_to_group_id) = get_wrapped_layout(object_layout);
+        let actual_ptr = self.allocator.alloc_zeroed(actual_layout);
+        if actual_ptr.is_null() {
+            handle_alloc_error(actual_layout);
+        }
+
+        // SAFETY: We know that `actual_ptr` with offset is at least aligned enough for casting it to `*mut u8` as the layout for
+        // the allocation backing this pointer ensures the last field in the layout is `u8.
+        #[allow(clippy::cast_ptr_alignment)]
+        let group_id_ptr = actual_ptr.add(offset_to_group_id).cast::<u8>();
+        let object_size = object_layout.size();
+        if TRACK_ALLOCATIONS.load(Ordering::Relaxed) {
+            try_with_suspended_allocation_group(
+                #[inline(always)]
+                |group_id| {
+                    group_id_ptr.write(group_id.as_raw());
+                    self.tracer.trace_allocation(object_size, group_id);
+                },
+            );
+        }
+
+        actual_ptr
+    }
```

Comment:
I get that we're able to skip the group ID zero write here since the allocation is already zeroed, but I don't think that optimization is worth having an entirely duplicated implementation of `alloc`.

We should remove this and just rely on the default implementation.